flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/LogicalTypeJsonSerializer.java

```diff
@@ -109,349 +123,358 @@ public void serialize(
-            // the json to support it, and the new json plan is compatible with the previous one.
-            if (externalSer.isInternalInput()) {
-                throw new TableException(
-                        "ExternalSerializer with `isInternalInput=true` is not supported.");
-            }
-            DataType dataType = externalSer.getDataType();
-            boolean isMapView = DataViewUtils.isMapViewDataType(dataType);
-            boolean isListView = DataViewUtils.isListViewDataType(dataType);
-            if (isMapView || isListView) {
-                jsonGenerator.writeStartObject();
-                jsonGenerator.writeStringField(FIELD_NAME_TYPE_NAME, LogicalTypeRoot.RAW.name());
-                jsonGenerator.writeBooleanField(FIELD_NAME_NULLABLE, rawType.isNullable());
-                if (isMapView) {
-                    jsonGenerator.writeStringField(
-                            FIELD_NAME_DATA_VIEW_CLASS, MapView.class.getName());
-                    KeyValueDataType keyValueDataType =
-                            DataViewUtils.extractKeyValueDataTypeForMapView(dataType);
-                    serializeDataTypeForDataView(
-                            FIELD_NAME_KEY_TYPE,
-                            keyValueDataType.getKeyDataType(),
-                            jsonGenerator,
-                            serializerProvider);
-                    serializeDataTypeForDataView(
-                            FIELD_NAME_VALUE_TYPE,
-                            keyValueDataType.getValueDataType(),
-                            jsonGenerator,
-                            serializerProvider);
-                } else {
-                    jsonGenerator.writeStringField(
-                            FIELD_NAME_DATA_VIEW_CLASS, ListView.class.getName());
-                    DataType elementType =
-                            DataViewUtils.extractElementDataTypeForListView(dataType);
-                    serializeDataTypeForDataView(
-                            FIELD_NAME_ELEMENT_TYPE,
-                            elementType,
-                            jsonGenerator,
-                            serializerProvider);
-                }
-                jsonGenerator.writeEndObject();
-                return;
-            }
+    /**
+     * Checks whether the given type can be serialized as a compact string created from {@link
+     * LogicalType#asSerializableString()}.
+     */
+    private static class CompactSerializationChecker extends LogicalTypeDefaultVisitor<Boolean> {
+
+        private final boolean serializeCatalogObjects;
+
+        CompactSerializationChecker(boolean serializeCatalogObjects) {
+            this.serializeCatalogObjects = serializeCatalogObjects;
         }
 
-        jsonGenerator.writeObject(rawType.asSerializableString());
-    }
+        @Override
+        public Boolean visit(CharType charType) {
+            return charType.getLength() > 0;
```

Comment:
Isn't this always true because CharType of length 0 is invalid?

Reply:
Unfortunately not. It is invalid in declarations but valid within a plan if you have something like `SELECT ""`.